### PR TITLE
FlatData: Handle multi-date formats

### PR DESCRIPTION
**Pull request by aziemchawdhary-gs**
## Summary

Modify Flatdata model to allow users to specify multiple data formats for Data columns

## How did you test this change?

Manual testing - the execution of this is internally used and have tested with test and user projects

### CODE REVIEW

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_FlatDataStoreBuilderHelper.ts [hunk 1]
```diff
@@ -169,9 +169,11 @@ const buildFlatDataProperty = (
     `Flat-data property 'value' field is missing`,
   );
   assertTrue(
-    isString(property.value) ||
-      isNumber(property.value) ||
-      isBoolean(property.value),
+    property.value.reduce(
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_FlatDataStoreBuilderHelper.ts:172]: I think we can use `Array.every` here instead

File: packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataDataType.ts [hunk 2]
```diff
@@ -68,13 +68,13 @@ export class V1_FlatDataDecimal extends V1_FlatDataNumber implements Hashable {
...
 export class V1_FlatDataDate extends V1_FlatDataDataType implements Hashable {
-  dateFormat?: string | undefined;
+  dateFormat?: string[] | undefined;
   timeZone?: string | undefined;
 
   get hashCode(): string {
     return hashArray([
       CORE_HASH_STRUCTURE.FLAT_DATA_DATE,
-      this.dateFormat ?? '',
+      this.dateFormat === undefined ? '' : this.dateFormat.join(','),
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataDataType.ts:77]: I think feel free to use `hashArray` here

File: packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataProperty.ts [hunk 3]
```diff
@@ -19,13 +19,13 @@ import { type Hashable, hashArray } from '@finos/legend-shared';
...
   name!: string;
-  value!: boolean | string | number;
+  value!: (boolean | string | number)[];
 
   get hashCode(): string {
     return hashArray([
       CORE_HASH_STRUCTURE.FLAT_DATA_PROPERTY,
       this.name,
-      this.value.toString(),
+      this.value.map((v) => v.toString()).join(','),
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataProperty.ts:28]: Let's just update to use `hashArray` for these places

**Review by akphi — CHANGES_REQUESTED** [review]
Let's use `hashArray` instead for hash computation. and please correct the failed pipeline checks. Thanks!

File: packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataProperty.ts (hunk 3 above)
Comment (aziemchawdhary-gs) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataProperty.ts:28]: this.value has type (boolean | string | number)[] but hashArray is defined on (string | Hashable)[] 

Do you want me to change hashArray?

Comment (akphi) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataProperty.ts:28]: Yeah and you can call toString on them, I prefer consistency and I think joining strings by , introduce edge cases

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_StoreSerializationHelper.ts [hunk 4]
```diff
@@ -92,19 +92,19 @@ const V1_flatDataDecimalSchema = createModelSchema(V1_FlatDataDecimal, {
...
-  dateFormat: optional(primitive()),
+  dateFormat: optional(list(primitive())),
   timeZone: optional(primitive()),
 });
 const V1_flatDataStrictDateSchema = createModelSchema(V1_FlatDataStrictDate, {
   _type: usingConstantValueSchema(
     V1_FlatDataDataTypeType.FLAT_DATA_STRICT_DATE,
   ),
-  dateFormat: optional(primitive()),
+  dateFormat: optional(list(primitive())),
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_StoreSerializationHelper.ts:107]: json -> v1 -> metamodel -> (transformer)  v1 -> json

**Review by akphi — CHANGES_REQUESTED** [review]

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_FlatDataStoreBuilderHelper.ts (hunk 1 above)
Comment (aziemchawdhary-gs) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_FlatDataStoreBuilderHelper.ts]: fixed

File: packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataDataType.ts (hunk 2 above)
Comment (aziemchawdhary-gs) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataDataType.ts]: fixed

File: packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataProperty.ts (hunk 3 above)
Comment (aziemchawdhary-gs) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataProperty.ts]: done as suggested

File: packages/legend-graph/src/graph/metamodel/pure/packageableElements/store/flatData/model/FlatDataDataType.ts [hunk 5]
```diff
@@ -79,7 +79,7 @@ export class FlatDataDate extends FlatDataDataType implements Hashable {
   get hashCode(): string {
     return hashArray([
       CORE_HASH_STRUCTURE.FLAT_DATA_DATE,
-      this.dateFormat == undefined ? '' : hashArray(this.dateFormat),
+      this.dateFormat === undefined ? '' : hashArray(this.dateFormat),
```
[the remark this answers is not in the document]
Comment (aziemchawdhary-gs) [reply packages/legend-graph/src/graph/metamodel/pure/packageableElements/store/flatData/model/FlatDataDataType.ts]: makes sense - fixed

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_StoreSerializationHelper.ts [hunk 6]
```diff
@@ -92,19 +92,63 @@ const V1_flatDataDecimalSchema = createModelSchema(V1_FlatDataDecimal, {
 });
 const V1_flatDataDateSchema = createModelSchema(V1_FlatDataDate, {
   _type: usingConstantValueSchema(V1_FlatDataDataTypeType.FLAT_DATA_DATE),
-  dateFormat: optional(list(primitive())),
+  // dateFormat : optional(list(primitive())),
```
[the remark this answers is not in the document]
Comment (aziemchawdhary-gs) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_StoreSerializationHelper.ts]: done

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_StoreSerializationHelper.ts [hunk 7]
```diff
@@ -92,19 +92,63 @@ const V1_flatDataDecimalSchema = createModelSchema(V1_FlatDataDecimal, {
...
+        : [val],
+  ),
   timeZone: optional(primitive()),
 });
 const V1_flatDataStrictDateSchema = createModelSchema(V1_FlatDataStrictDate, {
   _type: usingConstantValueSchema(
     V1_FlatDataDataTypeType.FLAT_DATA_STRICT_DATE,
   ),
-  dateFormat: optional(list(primitive())),
+  // dateFormat : optional(list(primitive())),
```
[the remark this answers is not in the document]
Comment (aziemchawdhary-gs) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_StoreSerializationHelper.ts]: done

**Comment by aziemchawdhary-gs** [comment]
I've changed the test data used in the flatdata roundtrip tests to have array values for FlatData properties and Data columns. 
This means that this change needs to be sync'd and released together with an (internal) engine release

File: packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataProperty.ts [hunk 8]
```diff
@@ -19,13 +19,13 @@ import { type Hashable, hashArray } from '@finos/legend-shared';
 
 export class V1_FlatDataProperty implements Hashable {
   name!: string;
-  value!: boolean | string | number;
+  value!: (boolean | string | number)[];
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/model/packageableElements/store/flatData/model/V1_FlatDataProperty.ts:22]: We can also create a type `V1_FlatDataPropertyValue = boolean | string | number` here

**Review by akphi — APPROVED** [review]

**Pull request merged**